social_protection/services.py

```diff
@@ -163,7 +243,74 @@ def _trigger_workflow(self,
         })
         upload.status = IndividualDataSourceUpload.Status.TRIGGERED
         upload.save(username=self.user.login_name)
+
+    def __save_validation_error_in_data_source(self, row, field_validation):
+        error_fields = []
+        for key, value in field_validation['validations'].items():
+            if not value['success']:
+                error_fields.append({
+                    "field_name": value['field_name'],
+                    "note": value['note']
+                })
+        individual_data_source = IndividualDataSource.objects.get(id=row['id'])
+        validation_column = {'validation_errors': error_fields}
+        individual_data_source.validations = validation_column
+        individual_data_source.save(username=self.user.username)
+
+    @register_service_signal('validation.create_task')
+    def _create_import_valid_items_task(self, benefit_plan, upload_id, user):
+        from social_protection.apps import SocialProtectionConfig
+        from tasks_management.services import TaskService
+        from tasks_management.apps import TasksManagementConfig
+        from tasks_management.models import Task
+        upload_record = BenefitPlanDataUploadRecords.objects.get(
+            data_upload_id=upload_id,
+            benefit_plan=benefit_plan,
+            is_deleted=False
+        )
+        json_ext = {
+            'benefit_plan_code': benefit_plan.code,
+            'source_name': upload_record.data_upload.source_name,
+            'workflow': upload_record.workflow,
+            'percentage_of_invalid_items': self.__calculate_percentage_of_invalid_items(upload_id),
+        }
+        TaskService(user).create({
+            'source': 'import_valid_items',
+            'entity': upload_record,
+            'status': Task.Status.RECEIVED,
+            'executor_action_event': TasksManagementConfig.default_executor_event,
+            'business_event': SocialProtectionConfig.validation_import_valid_items,
+            'json_ext': json_ext
+        })
+
+    def __fetch_summary_of_broken_items(self, upload_id):
+        return list(IndividualDataSource.objects.filter(
+            Q(is_deleted=False) &
+            Q(upload_id=upload_id) &
+            ~Q(validations__validation_errors=[])
+        ).values_list('uuid', flat=True))
+
+    def __fetch_summary_of_valid_items(self, upload_id):
+        return list(IndividualDataSource.objects.filter(
+            Q(is_deleted=False) &
+            Q(upload_id=upload_id) &
+            Q(validations__validation_errors=[])
+        ).values_list('uuid', flat=True))
+
+    def __calculate_percentage_of_invalid_items(self, upload_id):
+        total_items = len(self.__fetch_summary_of_broken_items(upload_id)) + len(
+            self.__fetch_summary_of_valid_items(upload_id))
+
+        if total_items == 0:
+            percentage_of_invalid_items = 0
+        else:
+            invalid_items = len(self.__fetch_summary_of_broken_items(upload_id))
```

Comment:
Your calling `.__fetch_summary_of_broken_items` twice in this method. Could you do it once in order to improve performance?